core/authority_discovery/publisher/address_publisher.hpp

```diff
@@ -0,0 +1,66 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_ADDRESS_PUBLISHER_IMPL_HPP
+#define KAGOME_ADDRESS_PUBLISHER_IMPL_HPP
+
+#include "application/app_state_manager.hpp"
+#include "blockchain/block_tree.hpp"
+#include "crypto/crypto_store/session_keys.hpp"
+#include "crypto/ed25519_provider.hpp"
+#include "crypto/sr25519_provider.hpp"
+#include "log/logger.hpp"
+#include "runtime/runtime_api/authority_discovery_api.hpp"
+
+#include <libp2p/crypto/key_marshaller.hpp>
+#include <libp2p/host/host.hpp>
+#include <libp2p/protocol/kademlia/kademlia.hpp>
+#include <memory>
+
+namespace kagome::authority_discovery {
+  class AddressPublisher
```

Comment:
Please comment.

Reply:
Also, I don't think I saw tests for it? It contains a pretty complex logic with various corner-cases.